common/src/main/scala/common/util/IntrospectableLazy.scala

```diff
@@ -0,0 +1,30 @@
+package common.util
+
+import scala.language.implicitConversions
+
+// Adapted from https://stackoverflow.com/a/17057490/818054
+
+object IntrospectableLazy {
+
+  def lazily[A](f: => A): IntrospectableLazy[A] = new IntrospectableLazy(f)
+
+  implicit def evalLazy[A](l: IntrospectableLazy[A]): A = l()
+
+}
+
+class IntrospectableLazy[A] private(f: => A) {
+
+  private var option: Option[A] = None
+
+  def apply(): A = {
+    this.synchronized {
```

Comment:
I do think it'll be nice to optimize for the (by far most common) common case by putting this `synchronized` only around the `None` case (and re-check whether it's still `None` inside the `synchronized`)